src/WebJobs.Script/Host/WorkerFunctionMetadataProvider.cs

```diff
@@ -7,90 +7,108 @@
+
+        public WorkerFunctionMetadataProvider(
+            IOptions<ScriptJobHostOptions> scriptOptions,
+            ILogger<WorkerFunctionMetadataProvider> logger,
+            IEnvironment environment,
+            IWebHostRpcWorkerChannelManager webHostRpcWorkerChannelManager)
         {
-            _logger = logger;
-            _dispatcher = invocationDispatcher;
-            _hostFunctionMetadataProvider = hostFunctionMetadataProvider;
             _scriptOptions = scriptOptions;
+            _logger = logger;
+            _environment = environment;
+            _channelManager = webHostRpcWorkerChannelManager;
+            _workerRuntime = _environment.GetEnvironmentVariable(EnvironmentSettingNames.FunctionWorkerRuntime);
         }
 
         public ImmutableDictionary<string, ImmutableArray<string>> FunctionErrors
            => _functionErrors.ToImmutableDictionary(kvp => kvp.Key, kvp => kvp.Value.ToImmutableArray());
 
-        public async Task<ImmutableArray<FunctionMetadata>> GetFunctionMetadataAsync(IEnumerable<RpcWorkerConfig> workerConfigs, IEnvironment environment, bool forceRefresh)
+        public async Task<FunctionMetadataResult> GetFunctionMetadataAsync(IEnumerable<RpcWorkerConfig> workerConfigs, IEnvironment environment, bool forceRefresh)
         {
             IEnumerable<FunctionMetadata> functions = new List<FunctionMetadata>();
             _logger.FunctionMetadataProviderParsingFunctions();
 
             if (_functions.IsDefaultOrEmpty || forceRefresh)
             {
                 IEnumerable<RawFunctionMetadata> rawFunctions = new List<RawFunctionMetadata>();
-                bool workerIndexing = Utility.CanWorkerIndex(workerConfigs, environment);
 
-                if (workerIndexing)
+                if (_channelManager == null)
                 {
-                    if (_dispatcher == null)
-                    {
-                        throw new InvalidOperationException(nameof(_dispatcher));
-                    }
-
-                    // start up GRPC channels
-                    await _dispatcher.InitializeAsync(new List<FunctionMetadata>());
+                    throw new InvalidOperationException(nameof(_channelManager));
+                }
 
-                    // get function metadata from worker, then validate it
-                    rawFunctions = await _dispatcher.GetWorkerMetadata();
+                // start up GRPC channels
+                await _channelManager.InitializeChannelAsync(_workerRuntime);
 
-                    if (IsDefaultIndexingRequired(rawFunctions))
-                    {
-                        _logger.LogDebug("Fallback to host indexing as worker denied indexing");
-                        functions = await _hostFunctionMetadataProvider.GetFunctionMetadataAsync(workerConfigs, environment, forceRefresh);
-                    }
-                    else if (!IsNullOrEmpty(rawFunctions))
+                var channels = _channelManager.GetChannels(_workerRuntime);
```

Comment:
At this point, do we always have 1 channel in `var channels`? or can there be scenarios with more than 1 channel here?

Reply:
I don't think we have scenarios where we expect more than 1 channel here. For our scenario we need atleast 1 channel.